Loosen coupling of `CollectionInputFilter` with `NotEmpty` validator

ramchale opened the pull request:
- Make getNotEmptyValidator private and add public setIsRequiredValidationMessage method
- Remove deprecated getOption method call in prepareRequiredValidationFailureMessage

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | yes
| QA            | yes

### Description

The call to `$notEmptyValidator->getOption('messageTemplates')` in CollectionInputFilter will need replacing to enable moving to ServiceManager v4. We could replace this with a call to `getMessageTemplates` but this is also deprecated. A direct call to the validator with an invalid value should achive the same results without CollectionInputFilter needing to understand the internals of the NotEmpty validator.

ramchale commented:
Just to check. Can `setNotEmptyValidator` and `getNotEmptyValidator' be removed from the public interface of `CollectionInputFilter` in v3, or is there use cases that I'm unaware of that require these to be public?

alexmerlin commented:
> Just to check. Can `setNotEmptyValidator` and `getNotEmptyValidator' be removed from the public interface of `CollectionInputFilter` in v3, or is there use cases that I'm unaware of that require these to be public?

@ramchale IMO they can stay private inside of `CollectionInputFilter`.
I'm not aware of any use case in related packages (checked in `laminas-form`, `laminas-input` and `laminas-validator`).
Maybe @gsteel can confirm?

gsteel commented on src/CollectionInputFilter.php:
```diff
@@ -85,17 +85,25 @@ public function getInputFilter()
+    /**
+     * Set a custom error message for the collection being empty.
+     * If not called, CollectionInputFilter will default to the NotEmpty validators IS_EMPTY message
+     */
```
```suggestion
     * @param non-empty-string $message
     */
```

ramchale replied:
Thanks

gsteel commented on src/CollectionInputFilter.php:
```diff
@@ -343,15 +334,7 @@ public function getUnknown()
-                : $message,
-        ];
+        $notEmptyValidator(null);
+        return $notEmptyValidator->getMessages();
```
The problem with this is now it is impossible to translate the error message, because it is not even possible to fetch the `NotEmpty` instance in order to set the validator _(I don't think that is possible in v3 anyhow)_.

Might it be better to, in `getNotEmptyValidator()` call:

```php
$this->factory->getValidatorPluginManager()->build(
    NotEmpty::class,
   [
       'messages' => [
           NotEmpty::IS_EMPTY => $this->requiredValidationMessage,
       ],
   ],
);

```

…instead of `new NotEmpty()`.

Doing this will mean that if the user wishes to use a translator, via some kind of validator initialiser, that translator will be used when necessary.

Edit… Thinking about this some more…

The validator is not actually executed with the input data at any point, so its purpose is purely to extract a single error message.

I'd suggest:

- removing the `getNotEmptyValidator` method, along with `$this->notEmptyValidator` property.
- renaming this method to `private function getEmptyValidationErrorMessages(): array`
- Storing the configured "empty" error message in a property such as `private string|null $emptyErrorMessage = null`
- and changing the body of this method to something like:

```php
/** @return array<string, string> */
private function getEmptyValidationErrorMessages(): array
{
    $options = $this->emptyErrorMessage === null
        ? []
        : ['messages' => [NotEmpty::IS_EMPTY => $this->emptyErrorMessage]];

    $validator = $this->factory->getValidatorPluginManager()->build(NotEmpty::class, $options);
    
    $validator->isValid(null);
    
    return $validator->getMessages();
}
```

ramchale replied:
Cool. Sounds sensible

gsteel commented on src/Input.php:
```diff
@@ -463,8 +463,9 @@ protected function prepareRequiredValidationFailureMessage()
         /** @psalm-var array<string, string> $templates */
-        $templates  = $notEmpty->getOption('messageTemplates');
-        $message    = $templates[NotEmpty::IS_EMPTY];
+        $templates = $notEmpty->getOption('messageTemplates');
+        $message   = $templates[NotEmpty::IS_EMPTY];
+        /** @psalm-suppress DeprecatedInterface */
```
The entire body of this method needs to be reconsidered.

```
$validator = $chain->hasANotEmptyValidator()
    ? $chain->getTheNotEmptyValidator()
    : $this->validators->makeADefaultNotEmtpy();

$validator->isValid(null);

return $validator->getMessages();
```

Then we can drop all the mucking about with translators and internal options inside `NotEmpty` that we have no business mucking about with.

ramchale replied:
👍 I can pick this up as a seperate change

gsteel commented on test/CollectionInputFilterTest.php:
```diff
@@ -746,32 +747,10 @@ public function testDuplicatedErrorMessages(): void
+        self::assertInstanceOf(
+            NotEmpty::class,
+            (new ReflectionMethod($this->inputFilter, 'getNotEmptyValidator'))->invoke($this->inputFilter)
+        );
```
I think we need to test here that a collection built via factory with…

`$factory->createInputFilter($configToMakeACollection)`

will yield the expected error message, when…

`$configToMakeACollection = ['required_message' => 'Whatever', ...];`

_and_

We get a non-empty error message when no custom error message has been set.

_and_

The setter for `setEmptyMessage()` or whatever the method is, affects the error message appropriately as per factory-backed construction.

We should avoid referring to any part of the NotEmpty validator in tests here.

gsteel reviewed (changes requested):
Thanks @ramchale - Sorry it took me a while to get to reviewing this!

ramchale commented:
@gsteel thanks for the suggestions. Let me know if I've missed anything.
If you think this is ready, it might be worth merging 2.34.x into 3.0.x and rebasing and squashing this PR to keep the history slightly cleaner

gsteel reviewed (approved):
Thanks @ramchale LGTM 👍 

I agree about a merge up of 2.34.x - I'll do that now so you can rebase here when you're ready

ramchale committed:
Loosen coupling of CollectionInputFilter with NotEmpty validator

- Remove CollectionInputFilter->getNotEmptyValidator and setNotEmptyValidator
- Add public setIsRequiredValidationMessage method
- Replace prepareRequiredValidationFailureMessage with getEmptyValidationErrorMessages (also removes deprecated getOption method call)
- Add Factory->getValidatorPluginManager

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale commented:
@gsteel thanks. That's rebased and ready for re-review
